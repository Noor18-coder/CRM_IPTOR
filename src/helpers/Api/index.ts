export * from './ApiRequest';
export * from './User';
export * from './CompanyInfo';
export * from './CurrencyInfo';
export * from './StagesInfo';
export * from './OpportunityType';
export * from './Attributes';
