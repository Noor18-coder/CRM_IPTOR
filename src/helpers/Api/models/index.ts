export * from './UserItem';
export * from './CompanyInfo';
export * from './ApiRequestControls';
export * from './Opportunity';
export * from './OpportunityDetails';
export * from './Users';
export * from './StageInfo';
export * from './Contacts';
export * from './Product';
export * from './Customer';
export * from './OpportunityTypes';
export * from './CustomerDetails';
export * from './AddOpportunity';
export * from './Items';
export * from './OpportunityAttributes';
export * from './Currency';
export * from './Attributes';
export * from './AddCustomer';
export * from './Countries';
export * from './ForeCastInfo';
