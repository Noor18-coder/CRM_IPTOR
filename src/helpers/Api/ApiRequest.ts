/**
 * Api Request Object
 */
import { ApiRequestControls } from './models';

export class ApiRequest<T> {
  IptorAPI: string;

  id: string;

  constructor(public method: string, public params?: T, public control?: ApiRequestControls) {
    this.IptorAPI = process.env.REACT_APP_IPTOR_API_VERSION ?? '';
    this.id = process.env.REACT_APP_IPTOR_API_ID ?? '';
  }
}
