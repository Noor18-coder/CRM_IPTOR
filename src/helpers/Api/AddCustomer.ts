import axios, { AxiosResponse } from 'axios';
import { get } from 'lodash';
import { AddBusinessPartnerDefaultParams, AddBusinessPartnerResponse, SaveUserDefinedFieldParam } from './models';
import { ApiRequest } from './ApiRequest';

export default class AddCustomerApi {
  /** API Method */
  private static apiMethod = 'crmBusinessPartner.add';

  private static updateApiMethod = 'crmBusinessPartner.update';

  private static saveAttributeValueMethod = 'mopAttribute.add';

  private static saveAttributeValueMethodFile = 'SRONAM';

  /**
   * Helper function to fetch Business Partner Info
   * API Method: crmBusinessPartner.add
   * */
  static async add(businessPartner: AddBusinessPartnerDefaultParams): Promise<AddBusinessPartnerResponse> {
    const requestData = new ApiRequest<AddBusinessPartnerDefaultParams>(this.apiMethod, businessPartner);
    const response = await axios.post<AxiosResponse>('/api/service', requestData);
    return get(response, 'data');
  }

  static async update(businessPartner: AddBusinessPartnerDefaultParams): Promise<AddBusinessPartnerResponse> {
    const requestData = new ApiRequest<AddBusinessPartnerDefaultParams>(this.updateApiMethod, businessPartner);
    const response = await axios.post<AxiosResponse>('/api/service', requestData);
    return get(response, 'data');
  }

  static async addAttributes(businessPartnerId: string, attributeType: string, attributeValue: string | number): Promise<AddBusinessPartnerResponse> {
    const params: SaveUserDefinedFieldParam = {
      attributeType,
      parentFile: this.saveAttributeValueMethodFile,
      parentId: businessPartnerId,
      attributeValue,
    };
    const requestData = new ApiRequest<SaveUserDefinedFieldParam>(this.saveAttributeValueMethod, params);
    const response = await axios.post<AxiosResponse>('/api/service', requestData);
    return get(response, 'data');
  }
}
