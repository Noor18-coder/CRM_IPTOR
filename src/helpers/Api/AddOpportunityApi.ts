import axios, { AxiosResponse } from 'axios';
import { AddOpportunityDefaultParams, AddOpportunityResponse, SaveUserDefinedFieldParam, AddItemToOpportunityParams, AddItemToOpportunityResponse, AddCustomerContactParams, AddCustomerContactRequestParams } from './models';
import { ApiRequest } from './ApiRequest';
import { get } from 'lodash';

export default class AddOpportunityApi {
  /** API Method */
  private static apiMethod: string = 'mopOpportunity.add';
  private static saveAttributeValueMethod: string = 'mopAttribute.add';
  private static saveAttributeValueMethodFile: string = 'SROMOPH';
  private static saveItemMethod: string = 'mopItem.add';
  private static saveContactFileName: string = 'SROMOPCH';
  private static saveContactMethod: string = 'mopContact.add';

  /**
   * Helper function to fetch Business Partner Info
   * API Method: mopOpportunity.get
   * @param freeTextSearch search term
   * 
   */
  static async add(opportunity: AddOpportunityDefaultParams): Promise<AddOpportunityResponse> {
    const requestData = new ApiRequest<AddOpportunityDefaultParams>(this.apiMethod, opportunity);
    const response = await axios.post<AxiosResponse>('/api/service', requestData);
    return get(response, 'data');
  }

  /**
   * Helper function to fetch Business Partner Info
   * API Method: mopOpportunity.get
   * @param freeTextSearch search term
   * 
   */
  static async addAttributes(opportunityId: string, attributeType: string, attributeValue: string | number): Promise<AddOpportunityResponse> {
    const params: SaveUserDefinedFieldParam = {
      attributeType: attributeType,
      parentFile: this.saveAttributeValueMethodFile,
      parentId: opportunityId,
      attributeValue: attributeValue
    }
    const requestData = new ApiRequest<SaveUserDefinedFieldParam>(this.saveAttributeValueMethod, params);
    const response = await axios.post<AxiosResponse>('/api/service', requestData);
    return get(response, 'data');
  }

  /**
   * Helper function to fetch Business Partner Info
   * API Method: mopOpportunity.get
   * @param freeTextSearch search term
   * 
   */
  static async addItem(opportunityId: string, item: string, quantity: number, unit:string): Promise<AddItemToOpportunityResponse> {
    const params: AddItemToOpportunityParams = {
      parentFile: this.saveAttributeValueMethodFile,
      parentId: opportunityId,
      item: item,
      quantity: quantity,
      unit:unit
    }
    const requestData = new ApiRequest<AddItemToOpportunityParams>(this.saveItemMethod, params);
    const response = await axios.post<AddItemToOpportunityResponse>('/api/service', requestData);
    return get(response, 'data');
  }

  /**
   * Helper function to fetch Business Partner Info
   * API Method: mopOpportunity.get
   * @param freeTextSearch search term
   * 
   */
   static async addContacts(params:AddCustomerContactParams): Promise<AxiosResponse> {
    const request:AddCustomerContactRequestParams  = {
      contactParentFile:this.saveAttributeValueMethodFile,
      contactParentId:params.contactPArentId,
      contactPerson:params.contactPerson,
      phone:params.phone,
      mobile:params.mobile,
      contactDC:'',
      whatsApp:'',
      linkedin:'',
      fax:'',
      email:params.email
    }

    const requestData = new ApiRequest<AddCustomerContactRequestParams>(this.saveContactMethod, request);
    const response = await axios.post<AxiosResponse>('/api/service', requestData);
    console.log(response);
    return get(response, 'data');
  }
}
