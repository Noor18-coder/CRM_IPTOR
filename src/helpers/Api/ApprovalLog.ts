import axios, { AxiosResponse } from 'axios';
import { get } from 'lodash';
import { ApprovalLogAddRequestParams, ApprovalLogAddResponse, DeleteApprovalLogsParams } from './models';
import { ApiRequest } from './ApiRequest';

export class ApprovalLog {
  /** API Method */
  private static apiMethod = 'crmOpportunityApprovalLog.add';

  private static deleteApprovalLogs = 'crmOpportunityApprovalLog.delete';

  /**
   * Helper function to fetch sales order list
   * @param params user id
   * @returns information about user
   */
  static async addApprovalLog(params: ApprovalLogAddRequestParams): Promise<ApprovalLogAddResponse> {
    const requestData = new ApiRequest<ApprovalLogAddRequestParams>(this.apiMethod, params);
    const response = await axios.post<ApprovalLogAddResponse>('/api/service', requestData);
    return get<AxiosResponse<ApprovalLogAddResponse>, 'data'>(response, 'data');
  }

  static async deleteOpportunityApprovalLogs(params: DeleteApprovalLogsParams): Promise<ApprovalLogAddResponse> {
    const requestData = new ApiRequest<DeleteApprovalLogsParams>(this.deleteApprovalLogs, params);
    const response = await axios.post<ApprovalLogAddResponse>('/api/service', requestData);
    return get(response, 'data.data.items', {});
  }
}
