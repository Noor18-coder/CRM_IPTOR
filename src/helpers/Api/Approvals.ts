import axios, { AxiosResponse } from 'axios';
import { get } from 'lodash';
import * as apiModels from './models';
import { ApiRequest } from './ApiRequest';

export class ApprovalInfo {
  /** API Method */
  private static allApprovalsMethod = 'crmOpportunityApprovalLogs.get';

  private static allNotesMethod = 'mopNotes.get';

  private static notesParentFile = 'SROMOPAL';

  static async getApprovalLogs(opportunityId: string): Promise<apiModels.ApprovalLogsDefault[]> {
    const requestData = new ApiRequest<apiModels.ApprovalLogsParams>(this.allApprovalsMethod, { opportunityId });
    const response = await axios.post<apiModels.ApprovalLogsResponse>('/api/service', requestData);
    return get<AxiosResponse<apiModels.ApprovalLogsResponse>, 'data', 'data', 'items', apiModels.ApprovalLogsDefault[]>(
      response,
      ['data', 'data', 'items'],
      []
    );
  }

  static async getAllNotes(parentId: string): Promise<apiModels.AllNotesDefault> {
    const requestData = new ApiRequest<apiModels.AllNotesParams>(this.allNotesMethod, { parentFile: this.notesParentFile, parentId });
    const response = await axios.post<apiModels.AllNotesResponse>('/api/service', requestData);
    const defaultData = {} as apiModels.AllNotesDefault;
    return get<AxiosResponse<apiModels.AllNotesResponse>, 'data', 'data', 'items', apiModels.AllNotesDefault>(
      response,
      ['data', 'data', 'items'],
      defaultData
    );
  }
}
