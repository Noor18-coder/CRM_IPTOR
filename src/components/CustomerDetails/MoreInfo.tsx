import React from 'react';
import { Card, Accordion, Image } from 'react-bootstrap';
import { Dispatch } from 'redux';
import { useDispatch, useSelector } from 'react-redux';
import _ from 'lodash';
import { OpportunityMoreInfoSection, OpportunityDetailsBasicInfo, CustomerDetailsDefault } from '../../helpers/Api/models';
import ImageConfig from '../../config/ImageConfig';
import { setBusinessPartnerWindowActive, setBusinessPartnerWindowGroup } from '../../store/AddCustomer/Actions';
import { getDateInFormat } from '../../helpers/utilities/lib';
import { AppState } from '../../store/store';

interface Props {
  title: string;
  data: OpportunityDetailsBasicInfo[];
  customerDetails: CustomerDetailsDefault;
}

export const MoreInfoAccordian: React.FC<OpportunityMoreInfoSection> = ({ title, data, customerDetails }) => {
  const keys = Object.keys(data);
  const [activeClass, setActiveClass] = React.useState('');

  const toggleAccordion = () => {
    setActiveClass(activeClass === '' ? 'active' : '');
  };

  return (
    <Accordion defaultActiveKey="0">
      <Card>
        <Accordion.Toggle className={activeClass} onClick={toggleAccordion} as={Card.Link} eventKey="1">
          {title}
        </Accordion.Toggle>
        <Accordion.Collapse eventKey="1">
          <>
            {keys.map((key: string) => {
              return <DisplayGroup title={key} data={data[key]} customerDetails={customerDetails} />;
            })}
          </>
        </Accordion.Collapse>
      </Card>
    </Accordion>
  );
};

export const DisplayGroup: React.FC<Props> = ({ title, data, customerDetails }) => {
  const state: AppState = useSelector((CustomerState: AppState) => CustomerState);
  const dispatch: Dispatch<any> = useDispatch();

  const filteredData = _.uniqBy(data, 'attributeType');
  const getUserName = (str: string) => {
    const userObj = state.users.users.find((obj) => obj.user === str);
    return userObj?.description ? userObj?.description : '--';
  };

  const getValue = (obj: any) => {
    const { customerAttributes } = state.enviornmentConfigs;

    const dataObject = data.find((object: any) => {
      return object.attributeType === obj.attributeType;
    });
    const field = customerAttributes.find((object: any) => {
      return object.attributeType === obj.attributeType;
    });
    const value = dataObject && dataObject.attributeValue && dataObject.attributeValue !== '' ? dataObject.attributeValue : '--';
    if (field?.valueFormatDesc === 'DATE') {
      if (value === '--') {
        return '--';
      } else {
        const year = value.substring(0, 4);
        const month = value.substring(4, 6);
        const day = value.substring(6, 8);
        const tempDate = `${year}-${month}-${day}`;
        return getDateInFormat(new Date(tempDate));
      }
    } else if (field?.valueFormatDesc === 'BOOLEAN') {
      return (
        <span className="checkbox-label">
          <label className="switch value-checkbox">
            <input type="checkbox" tabIndex={0} checked={value !== 'N'} />
            <span className="slider round disabled-checkbox">&nbsp;</span>
          </label>
        </span>
      );
    } else if (field?.attributeType === 'owner' || field?.attributeType === 'PS_ACCOUNT_OWNER') {
      return getUserName(obj.attributeValue);
    }
    return value;
  };

  const getMultipleValue = (obj: any) => {
    const { customerAttributes } = state.enviornmentConfigs;
    const field = customerAttributes.find((object: any) => {
      return object.attributeType === obj.attributeType;
    });

    const value = obj && obj.attributeValue && obj.attributeValue !== '' ? obj.attributeValue : '--';

    if (field?.valueFormatDesc === 'DATE') {
      if (value === '--') {
        return '--';
      } else {
        const year = value.substring(0, 4);
        const month = value.substring(4, 6);
        const day = value.substring(6, 8);
        const tempDate = `${year}-${month}-${day}`;
        return getDateInFormat(new Date(tempDate));
      }
    } else if (field?.attributeType === 'owner' || field?.attributeType === 'PS_ACCOUNT_OWNER') {
      return getUserName(obj.attributeValue);
    }
    return value;
  };

  const toggleDrawer = (open: boolean, groupType: string) => {
    dispatch(setBusinessPartnerWindowActive(true));
    dispatch(setBusinessPartnerWindowGroup(groupType));
  };

  return (
    <div className="more-info-group-container">
      <div className="more-info-group-name">
        <span>{title}</span>
        {!customerDetails.active ? (
          ((!!state.auth.user.role && state.auth.user.role === 'Admin') || state.auth.user.user === customerDetails.owner) && (
            <span className="group-icon">
              <Image src={ImageConfig.EDIT_ICON} className="action-icon" alt="Edit" title="Edit" onClick={() => toggleDrawer(true, title)} />
            </span>
          )
        ) : (
          <span className="group-icon">
            <Image src={ImageConfig.EDIT_ICON} className="action-icon" alt="Edit" title="Edit" onClick={() => toggleDrawer(true, title)} />
          </span>
        )}
      </div>
      <div className="accr-body-container">
        <ul className="list-inline bdy-list-item accr-list-columns">
          {filteredData.map((obj: any) => {
            const dataObj = data.filter((dataObject: any) => {
              return dataObject.attributeType === obj.attributeType;
            });
            if (dataObj.length > 1) {
              return dataObj.map((elem: any) => {
                return (
                  <li className="list-inline-item">
                    <span>{obj.description}</span>
                    {getMultipleValue(elem)}
                  </li>
                );
              });
            } else {
              return (
                <li className="list-inline-item">
                  <span>{obj.description}</span>
                  {getValue(obj)}
                </li>
              );
            }
          })}
        </ul>
      </div>
    </div>
  );
};

export const MoreInfoAccordianMobile: React.FC<OpportunityMoreInfoSection> = ({ title, data, customerDetails }) => {
  const state: AppState = useSelector((CustomerState: AppState) => CustomerState);
  const keys = Object.keys(data);
  const [activeClass, setActiveClass] = React.useState('');
  const dispatch: Dispatch<any> = useDispatch();

  const toggleAccordion = () => {
    setActiveClass(activeClass === '' ? 'active' : '');
  };

  const toggleDrawer = (open: boolean, groupType: string) => {
    document.body.classList.add('body-scroll-hidden');
    dispatch(setBusinessPartnerWindowActive(true));
    dispatch(setBusinessPartnerWindowGroup(groupType));
  };

  return (
    <div className="opp-moreinfo-sec">
      {title}
      <>
        {keys.map((key: string) => {
          return (
            <Accordion defaultActiveKey="0">
              <Accordion.Toggle className={activeClass} onClick={toggleAccordion} as={Card.Link} eventKey="1">
                <span className="edit-subtitle">{key}</span>
                {!customerDetails.active ? (
                  ((!!state.auth.user.role && state.auth.user.role === 'Admin') || state.auth.user.user === customerDetails.owner) && (
                    <button className="panel-close-icon link-anchor-button" onClick={() => toggleDrawer(true, key)} type="button">
                      <img src={ImageConfig.EDIT_ICON} className="mob-edit-icon-cust action-icon" alt="Edit" />
                    </button>
                  )
                ) : (
                  <button className="panel-close-icon link-anchor-button" onClick={() => toggleDrawer(true, key)} type="button">
                    <img src={ImageConfig.EDIT_ICON} className="mob-edit-icon-cust action-icon" alt="Edit" />
                  </button>
                )}
              </Accordion.Toggle>
              <Accordion.Collapse eventKey="1">
                <DisplayGroupMobile title={key} data={data[key]} customerDetails={customerDetails} />
              </Accordion.Collapse>
            </Accordion>
          );
        })}
      </>
    </div>
  );
};

export const DisplayGroupMobile: React.FC<Props> = ({ data }) => {
  const state: AppState = useSelector((CustomerState: AppState) => CustomerState);

  const getUserName = (str: string) => {
    const userObj = state.users.users.find((obj) => obj.user === str);
    return userObj?.description ? userObj?.description : '--';
  };

  const getValue = (obj: any) => {
    const { customerAttributes } = state.enviornmentConfigs;

    const dataObject = data.find((object: any) => {
      return object.attributeType === obj.attributeType;
    });
    const field = customerAttributes.find((object: any) => {
      return object.attributeType === obj.attributeType;
    });
    const value = dataObject && dataObject.attributeValue ? dataObject.attributeValue : '--';

    if (field?.valueFormatDesc === 'DATE') {
      if (value === '') {
        return '--';
      } else {
        const year = value.substring(0, 4);
        const month = value.substring(4, 6);
        const day = value.substring(6, 8);
        const tempDate = `${year}-${month}-${day}`;
        return getDateInFormat(new Date(tempDate));
      }
    } else if (field?.valueFormatDesc === 'BOOLEAN') {
      return (
        <span className="checkbox-label">
          <label className="switch value-checkbox">
            <input type="checkbox" tabIndex={0} checked={value !== 'N'} />
            <span className="slider round disabled-checkbox">&nbsp;</span>
          </label>
        </span>
      );
    } else if (field?.description === 'Account Owner') {
      return getUserName(obj.attributeValue);
    }
    return value;
  };

  return (
    <div className="more-info-group-container">
      <div className="accr-body-container">
        {data.map((obj: any) => {
          return (
            <ul className="list-inline bdy-list-item">
              <li className="list-inline-item">
                <span>{obj.description}</span>
                {getValue(obj)}
              </li>
            </ul>
          );
        })}
      </div>
    </div>
  );
};
