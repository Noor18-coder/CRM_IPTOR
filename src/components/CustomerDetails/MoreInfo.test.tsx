import { Provider } from 'react-redux';
import renderer from 'react-test-renderer';
import { MoreInfoAccordian, DisplayGroup, MoreInfoAccordianMobile, DisplayGroupMobile } from './MoreInfo';
import StoreMock from '../../mocks/Store.mock';
import { OpportunityDetailsBasicInfo, CustomerDetailsDefault } from '../../helpers/Api/models';

const store = StoreMock.createInitialStore();
const data = [] as OpportunityDetailsBasicInfo[];
const customerDetails = {} as CustomerDetailsDefault;

describe('[CustomerDetails] MoreInfoAccordian', () => {
  it('should renders correctly', () => {
    const tree = renderer
      .create(
        <Provider store={store}>
          <MoreInfoAccordian title="title" data={data} />
        </Provider>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});

describe('[CustomerDetails] DisplayGroup', () => {
  it('should renders correctly', () => {
    const tree = renderer
      .create(
        <Provider store={store}>
          <DisplayGroup title="title" data={data} customerDetails={customerDetails} />
        </Provider>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});

describe('[CustomerDetails] MoreInfoAccordianMobile', () => {
  it('should renders correctly', () => {
    const tree = renderer
      .create(
        <Provider store={store}>
          <MoreInfoAccordianMobile title="title" data={data} />
        </Provider>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});

describe('[CustomerDetails] DisplayGroupMobile', () => {
  it('should renders correctly', () => {
    const tree = renderer
      .create(
        <Provider store={store}>
          <DisplayGroupMobile title="title" data={data} customerDetails={customerDetails} />
        </Provider>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
