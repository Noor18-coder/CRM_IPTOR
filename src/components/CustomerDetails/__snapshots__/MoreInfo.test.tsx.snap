// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`[CustomerDetails] DisplayGroup should renders correctly 1`] = `
<div
  className="more-info-group-container"
>
  <div
    className="more-info-group-name"
  >
    <span>
      title
    </span>
  </div>
  <div
    className="accr-body-container"
  >
    <ul
      className="list-inline bdy-list-item accr-list-columns"
    />
  </div>
</div>
`;

exports[`[CustomerDetails] DisplayGroupMobile should renders correctly 1`] = `
<div
  className="more-info-group-container"
>
  <div
    className="accr-body-container"
  />
</div>
`;

exports[`[CustomerDetails] MoreInfoAccordian should renders correctly 1`] = `
<div
  className="accordion"
>
  <div
    className="card"
  >
    <a
      className="card-link"
      onClick={[Function]}
    >
      title
    </a>
    <div
      aria-expanded={null}
      className="collapse"
    />
  </div>
</div>
`;

exports[`[CustomerDetails] MoreInfoAccordianMobile should renders correctly 1`] = `
<div
  className="opp-moreinfo-sec"
>
  title
</div>
`;
